File: bench/csharp/Bench.cs
// Nine benchmark functions, one per bench/c file, same signatures.
// Run: Bench <name>; the JIT listing of that one method is taken with
// DOTNET_JitDisasm=<Method> DOTNET_TieredCompilation=0 (FullOpts).
using System;
using System.Runtime.CompilerServices;

static class Bench
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Fib(int n) { if (n <= 1) return n; return Fib(n - 1) + Fib(n - 2); }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Fact(int n) { if (n == 0) return 1; return (long)n * Fact(n - 1); }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Gcd(int a, int b) { while (b != 0) { int t = b; b = a % b; a = t; } return a; }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Sum(int n) { long s = 0; for (int i = 1; i <= n; i++) s += i; return s; }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Ack(int m, int n) { if (m == 0) return n + 1; if (n == 0) return Ack(m - 1, 1); return Ack(m - 1, (int)Ack(m, n - 1)); }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Tak(long x, long y, long z) { if (y >= x) return z; return Tak(Tak(x - 1, y, z), Tak(y - 1, z, x), Tak(z - 1, x, y)); }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Collatz(long n) { long steps = 0; while (n != 1) { if (n % 2 == 0) n = n / 2; else n = 3 * n + 1; steps++; } return steps; }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Compute(int n) { long acc = 0; for (int i = n; i > 0; i--) { long a = i + 1, b = i * 2, c = i - 3; long d = a * b, e = b + c, f = c * a; acc += d + e + f; } return acc; }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static long Regright(int n) { if (n <= 0) return 1; return Regright(n - 1) * n - n; }

    static int Main(string[] args)
    {
        string name = args.Length > 0 ? args[0] : "fib";
        switch (name)
        {
            case "fib": Console.WriteLine(Fib(35)); break;
            case "fact": Console.WriteLine(Fact(20)); break;
            case "gcd": Console.WriteLine(Gcd(46368, 28657)); break;
            case "sum": Console.WriteLine(Sum(1000000)); break;
            case "ack": Console.WriteLine(Ack(3, 9)); break;
            case "tak": Console.WriteLine(Tak(24, 16, 8)); break;
            case "collatz": Console.WriteLine(Collatz(837799)); break;
            case "locals": Console.WriteLine(Compute(1000)); break;
            case "regright": Console.WriteLine(Regright(12)); break;
            default: Console.Error.WriteLine("unknown bench " + name); return 2;
        }
        return 0;
    }
}

File: bench/fsharp/Bench.fs
// Nine benchmark functions, one per bench/c file, same signatures.
// Loops are written in the function body (for/while, or a self tail call the
// compiler turns into a loop), never as an inner `let rec go`, which F#
// compiles to a SEPARATE method that a listing of the outer name would miss.
// Run: Bench <name>; the JIT listing of that one function is taken with
// DOTNET_JitDisasm=<name> DOTNET_TieredCompilation=0 (FullOpts).
module Bench

open System.Runtime.CompilerServices

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec fib (n: int) : int64 = if n <= 1 then int64 n else fib (n - 1) + fib (n - 2)

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec fact (n: int) : int64 = if n = 0 then 1L else int64 n * fact (n - 1)

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec gcd (a: int) (b: int) : int = if b = 0 then a else gcd b (a % b)

[<MethodImpl(MethodImplOptions.NoInlining)>]
let sum (n: int) : int64 =
    let mutable s = 0L
    for i in 1 .. n do
        s <- s + int64 i
    s

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec ack (m: int) (n: int) : int64 =
    if m = 0 then int64 (n + 1)
    elif n = 0 then ack (m - 1) 1
    else ack (m - 1) (int (ack m (n - 1)))

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec tak (x: int64) (y: int64) (z: int64) : int64 =
    if y >= x then z else tak (tak (x - 1L) y z) (tak (y - 1L) z x) (tak (z - 1L) x y)

[<MethodImpl(MethodImplOptions.NoInlining)>]
let collatz (n0: int64) : int64 =
    let mutable n = n0
    let mutable steps = 0L
    while n <> 1L do
        if n % 2L = 0L then n <- n / 2L else n <- 3L * n + 1L
        steps <- steps + 1L
    steps

[<MethodImpl(MethodImplOptions.NoInlining)>]
let compute (n: int) : int64 =
    let mutable acc = 0L
    let mutable i = n
    while i > 0 do
        let a = int64 (i + 1)
        let b = int64 (i * 2)
        let c = int64 (i - 3)
        let d = a * b
        let e = b + c
        let f = c * a
        acc <- acc + d + e + f
        i <- i - 1
    acc

[<MethodImpl(MethodImplOptions.NoInlining)>]
let rec regright (n: int) : int64 = if n <= 0 then 1L else regright (n - 1) * int64 n - int64 n

[<EntryPoint>]
let main (args: string[]) : int =
    let name = if args.Length > 0 then args.[0] else "fib"
    match name with
    | "fib" -> printfn "%d" (fib 35); 0
    | "fact" -> printfn "%d" (fact 20); 0
    | "gcd" -> printfn "%d" (gcd 46368 28657); 0
    | "sum" -> printfn "%d" (sum 1000000); 0
    | "ack" -> printfn "%d" (ack 3 9); 0
    | "tak" -> printfn "%d" (tak 24L 16L 8L); 0
    | "collatz" -> printfn "%d" (collatz 837799L); 0
    | "locals" -> printfn "%d" (compute 1000); 0
    | "regright" -> printfn "%d" (regright 12); 0
    | _ -> eprintfn "unknown bench %s" name; 2